File: system/extras/simpleperf/utils_test.cpp
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>

#include "get_test_data.h"
#include "utils.h"

static bool ModulesMatch(const char* p, const char* q) {
  if (p == nullptr && q == nullptr) {
    return true;
  }
  if (p != nullptr && q != nullptr) {
    return strcmp(p, q) == 0;
  }
  return false;
}

static bool KernelSymbolsMatch(const KernelSymbol& sym1,
                               const KernelSymbol& sym2) {
  return sym1.addr == sym2.addr && sym1.type == sym2.type &&
         strcmp(sym1.name, sym2.name) == 0 &&
         ModulesMatch(sym1.module, sym2.module);
}

TEST(utils, ProcessKernelSymbols) {
  std::string data =
      "ffffffffa005c4e4 d __warned.41698   [libsas]\n"
      "aaaaaaaaaaaaaaaa T _text\n"
      "cccccccccccccccc c ccccc\n";
  KernelSymbol expected_symbol;
  expected_symbol.addr = 0xffffffffa005c4e4ULL;
  expected_symbol.type = 'd';
  expected_symbol.name = "__warned.41698";
  expected_symbol.module = "libsas";
  ASSERT_TRUE(ProcessKernelSymbols(
      data,
      std::bind(&KernelSymbolsMatch, std::placeholders::_1, expected_symbol)));

  expected_symbol.addr = 0xaaaaaaaaaaaaaaaaULL;
  expected_symbol.type = 'T';
  expected_symbol.name = "_text";
  expected_symbol.module = nullptr;
  ASSERT_TRUE(ProcessKernelSymbols(
      data,
      std::bind(&KernelSymbolsMatch, std::placeholders::_1, expected_symbol)));

  expected_symbol.name = "non_existent_symbol";
  ASSERT_FALSE(ProcessKernelSymbols(
      data,
      std::bind(&KernelSymbolsMatch, std::placeholders::_1, expected_symbol)));
}

TEST(utils, ConvertBytesToValue) {
  char buf[8];
  for (int i = 0; i < 8; ++i) {
    buf[i] = i;
  }
  ASSERT_EQ(0x1ULL, ConvertBytesToValue(buf + 1, 1));
  ASSERT_EQ(0x201ULL, ConvertBytesToValue(buf + 1, 2));
  ASSERT_EQ(0x05040302ULL, ConvertBytesToValue(buf + 2, 4));
  ASSERT_EQ(0x0706050403020100ULL, ConvertBytesToValue(buf, 8));
}

TEST(utils, ArchiveHelper) {
  std::unique_ptr<ArchiveHelper> ahelper = ArchiveHelper::CreateInstance(GetTestData(APK_FILE));
  ASSERT_TRUE(ahelper);
  bool found = false;
  ZipEntry lib_entry;
  ASSERT_TRUE(ahelper->IterateEntries([&](ZipEntry& entry, const std::string& name) {
    if (name == NATIVELIB_IN_APK) {
      found = true;
      lib_entry = entry;
      return false;
    }
    return true;
  }));
  ASSERT_TRUE(found);
  ZipEntry entry;
  ASSERT_TRUE(ahelper->FindEntry(NATIVELIB_IN_APK, &entry));
  ASSERT_EQ(entry.offset, lib_entry.offset);
  std::vector<uint8_t> data;
  ASSERT_TRUE(ahelper->GetEntryData(entry, &data));

  // Check reading wrong file formats.
  ASSERT_FALSE(ArchiveHelper::CreateInstance(GetTestData(ELF_FILE)));
  ASSERT_FALSE(ArchiveHelper::CreateInstance("/dev/zero"));
}

File: system/timezone/tzlookup_generator/src/test/java/com/android/libcore/timezone/tzlookup/UtilsTest.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.libcore.timezone.tzlookup;

import org.junit.Test;

import java.text.ParseException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.android.libcore.timezone.tzlookup.Utils.parseUtcOffsetToMillis;
import static com.android.libcore.timezone.tzlookup.Utils.toUtcOffsetString;
import static java.util.concurrent.TimeUnit.HOURS;
import static java.util.concurrent.TimeUnit.MINUTES;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class UtilsTest {

    @Test
    public void subtract() {
        Set<String> a = set("Foo", "Bar");
        Set<String> b = set("Bar", "Baz");
        Set<String> empty = set();
        assertEquals(set("Foo"), Utils.subtract(a, b));
        assertEquals(set("Baz"), Utils.subtract(b, a));
        assertEquals(a, Utils.subtract(a, empty));
        assertEquals(empty, Utils.subtract(empty, a));
    }

    @Test
    public void toUpperCase() {
        assertEquals(list("ONE", "TWO", "THREE"), Utils.toUpperCase(list("one", "Two", "THREE")));
        assertEquals(list(), Utils.toUpperCase(list()));
    }

    @Test
    public void allLowerCaseAscii() {
        assertTrue(Utils.allLowerCaseAscii(list("one", "a", "z")));
        assertFalse(Utils.allLowerCaseAscii(list("one", "a", "Z")));
        assertFalse(Utils.allLowerCaseAscii(list("one", "A", "z")));
        assertFalse(Utils.allLowerCaseAscii(list("onE", "a", "z")));
    }

    @Test
    public void utcOffsets() throws Exception {
        assertParseUtcFails("");
        assertParseUtcFails("0");
        assertParseUtcFails("0:0");

        assertEquals(0, parseUtcOffsetToMillis("0:00"));
        assertEquals(0, parseUtcOffsetToMillis("00:00"));
        assertEquals(0, parseUtcOffsetToMillis("-0:00"));
        assertParseUtcFails("+00:00");

        assertEquals("00:00", toUtcOffsetString(0));

        long hourInMillis = HOURS.toMillis(1);
        assertEquals(hourInMillis, parseUtcOffsetToMillis("1:00"));
        assertEquals(hourInMillis, parseUtcOffsetToMillis("01:00"));
        assertEquals(-hourInMillis, parseUtcOffsetToMillis("-1:00"));
        assertEquals(-hourInMillis, parseUtcOffsetToMillis("-01:00"));

        assertEquals("01:00", toUtcOffsetString(hourInMillis));
        assertEquals("-01:00", toUtcOffsetString(-hourInMillis));

        long hourAndAHalfInMillis = hourInMillis + MINUTES.toMillis(30);
        assertEquals(hourAndAHalfInMillis, parseUtcOffsetToMillis("1:30"));
        assertEquals(hourAndAHalfInMillis, parseUtcOffsetToMillis("01:30"));
        assertEquals(-hourAndAHalfInMillis, parseUtcOffsetToMillis("-1:30"));
        assertEquals(-hourAndAHalfInMillis, parseUtcOffsetToMillis("-01:30"));

        assertEquals("01:30", toUtcOffsetString(hourAndAHalfInMillis));
        assertEquals("-01:30", toUtcOffsetString(-hourAndAHalfInMillis));

        long hourAndTenInMillis = hourInMillis + MINUTES.toMillis(10);
        assertEquals(hourAndTenInMillis, parseUtcOffsetToMillis("01:10"));
        assertEquals(-hourAndTenInMillis, parseUtcOffsetToMillis("-01:10"));

        assertEquals("01:10", toUtcOffsetString(hourAndTenInMillis));
        assertEquals("-01:10", toUtcOffsetString(-hourAndTenInMillis));

        assertParseUtcFails("01:60");
        assertParseUtcFails("01:90");
        assertParseUtcFails("01:-1");
    }

    private void assertParseUtcFails(String s) {
        try {
            parseUtcOffsetToMillis(s);
            fail();
        } catch (ParseException expected) {

        }
    }

    @Test
    public void allUnique() throws Exception {
        assertTrue(Utils.allUnique(list()));
        assertTrue(Utils.allUnique(list("One")));
        assertTrue(Utils.allUnique(list("One", "Two")));
        assertFalse(Utils.allUnique(list("One", "Two", "One")));
    }

    @Test
    public void setEquals() throws Exception {
        assertTrue(Utils.setEquals(list("One", "Two"), list("Two", "One")));
        // Dupes allowed.
        assertTrue(Utils.setEquals(list("One", "One", "Two"), list("Two", "One")));

        assertFalse(Utils.setEquals(list("One", "Two", "Three"), list("Two", "One")));
    }

    private static <X> List<X> list(X... values) {
        return Arrays.asList(values);
    }

    private static <X> Set<X> set(X... values) {
        return new HashSet<>(Arrays.asList(values));
    }
}
